frame/staking/src/lib.rs

```diff
@@ -743,6 +759,55 @@ impl<T: Trait> SessionInterface<<T as frame_system::Trait>::AccountId> for T whe
 	}
 }
 
+pub mod weight {
+	use super::*;
+
+	/// All weight notes are pertaining to the case of a better solution, in which we execute
+	/// the longest code path.
+	/// Weight: 0 + (2.2 μs * v) + (2.4 μs * n) + (120 μs * a ) + (7.6 μs * w ) with:
+	/// * v validators in snapshot validators,
+	/// * n nominators in snapshot nominators,
+	/// * a assignment in the submitted solution
+	/// * w winners in the submitted solution
+	///
+	/// State reads:
+	/// 	- Initial checks:
+	/// 		- ElectionState, CurrentEra, QueuedScore
+	/// 		- SnapshotValidators.len()
+	/// 		- ValidatorCount
+	/// 		- SnapshotValidators
+	/// 		- SnapshotNominators
+	/// 	- Iterate over nominators:
+	/// 		- compact.len() * Nominators(who)
+	/// 		- (non_self_vote_edges) * SlashingSpans
+	/// 	- For `assignment_ratio_to_staked`: Basically read the staked value of each stash.
+	/// 		- (winners.len() + compact.len()) * (Ledger + Bonded)
+	/// 		- TotalIssuance (read a gzillion times potentially, but well it is cached.)
+	/// - State writes:
+	/// 	- QueuedElected, QueuedScore
+	pub fn weight_for_submit_solution<T: Trait>(
+		winners: &Vec<ValidatorIndex>,
+		compact: &CompactAssignments,
+		size: &ElectionSize,
+	) -> Weight {
+		(2_200 * WEIGHT_PER_NANOS).saturating_mul(size.validators as Weight)
+			.saturating_add((2_400 * WEIGHT_PER_NANOS).saturating_mul(size.nominators as Weight))
+			.saturating_add((120 * WEIGHT_PER_MICROS).saturating_mul(compact.len() as Weight))
+			.saturating_add((7_600 * WEIGHT_PER_NANOS).saturating_mul(winners.len() as Weight))
+			// Initial checks
+			.saturating_add(T::DbWeight::get().reads(7))
+			// Nominators
+			.saturating_add(T::DbWeight::get().reads(compact.len() as Weight))
+			// SlashingSpans (upper bound for invalid solution)
+			.saturating_add(T::DbWeight::get().reads(compact.edge_count() as Weight))
```

Comment:
I'm still trying to see how we could check nominations while (or before) querying slashspans so we don't have to weight more than necessary this number of read.
because edge_count can be `16*assignments` so we are increasing the number of read (bought in advance) by a factor of 10.

Maybe we can afford it I have to check what is the price of this extrinsic with real values. (if price is not too much then no need to precise it anymore, refund it all fine)

Reply:
I think wit the refund scheme that hopefully is correct [in the comment above](https://github.com/paritytech/substrate/pull/6032#discussion_r427789125), it should be fine.

Reply:
For information

in case we have all assignment with 16 nominations, so we will have:
* 16 (slashingspans)
* + 1 (Nominators)
* + 2 (assignment_to_ratio)
reads per assignment

so with our current formula we have `w=(19reads + 120us) * a + some_other_terms=(475+120us) *a + some_other_terms`
let's say some_other_terms are negligible in front of cost of `a`.
for an extrinsic of weight=1second we can fit `1s/595us=1700` assigment

on the other hand if we don't pay for the slashing spans in advance (if we refactor)
for an extrinsic of weight=1second we can fit `1s/(3reads+120us)=6900` assignment.

So I think for now we are fine.